docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -0,0 +1,519 @@
+
+Proposed revision
+-----------
+
+To address two limitations more fundamentally, this document proposes to
+use a single table that includes ECALLs or OCALLs in multiple EDL files.
+This section describes the detail of the design and considerations.
+
+**Maintain Single Table for ECALL and OCALL**
+
+Instead of adopting the per-EDL tables, maintaining a single table (i.e., one
+OCALL table and one ECALL table) provides finer granularity to keep track
+of ECALLs or OCALLs from multiple EDL files that are either imported or directly included.
+To avoid duplicated entries, each ECALL or OCALL is registered by its hash of the name.
+  
+**Add the registration functions of ECALLs and OCALLs in `_t.c` and `_u.c`**
+
+Instead of using wrappers for `_t.c` and `_u.c` files, each `_t.c` or `_u.c` now
+includes functions that register the list of OCALLs/ECALLs by passing the function
+pointers and the hashes of function names. See the following example.
+  
+```
+static oe_ecall_func_t __oe_ecalls_table[] = {
+    (oe_ecall_func_t) ecall_oe_verify_report_ecall,
+    (oe_ecall_func_t) ecall_oe_get_public_key_ecall,
+    (oe_ecall_func_t) ecall_oe_get_public_key_by_policy_ecall,
+    (oe_ecall_func_t) ecall_oe_log_init_ecall
+};
+
+static uint32_t __oe_ecalls_hash_list[] = {
+    3802944385 /* CRC32 of oe_verify_report_ecall. */,
+    1967956880 /* CRC32 of oe_get_public_key_ecall. */,
+    550431498 /* CRC32 of oe_get_public_key_by_policy_ecall. */,
+    4072068939 /* CRC32 of oe_log_init_ecall. */
+};
+
+static size_t __oe_ecalls_table_size = OE_COUNTOF(__oe_ecalls_table);
+
+/* Ecall table registration function. */
+static void register_core_ecall_functions()
+{
+    oe_register_ecall_functions(
+      __oe_ecalls_table,
+      __oe_ecalls_hash_list,
+      __oe_ecalls_table_size);
+}
+```
+
+Each table entry is then defined as the following struct.
+
+```
+typedef struct _oe_ocall_t {
+    oe_ocall_func_t ocall;
+    uint32_t hash;
+} oe_ocall_t;
+```
+
+**Maintain the List of Registration Functions**
+
+Each registration function in a EDL file is added to a list via the constructor.
+On Linux, we can do this via ` __attribute__((constructor))` as follows.
+We can use an equivalent approach for the constructor on Windows.
+```
+/* Constructor to register the callback. */
+ __attribute__((constructor)) static void core_ecall_functions_constructor()
+{
+    oe_ecall_init(register_core_ecall_functions);
+}
+```
+
+The list of registration functions can then be invoked during the host and enclave
+initialization.
+  
+**Cache the Function Id for the Const-Time Look-Up**
+
+One important property of current implementation is the const-time function look-up that
+this proposal would like to maintain. To this end, the design is obtaining the id by hash
+in the first invocation and caching the id for subsequent ones.
+
+- Caching an OCALL id.
+
+  Because an enclave only interacts with a single host, the case of OCALL wrapper is straightforward:
```

Comment:
Re "an enclave only interacts with a single host"... this is not true for OP-TEE in general which supports single-instanced enclaves.  As such, OCALLs are not as straightforward as this implies.   However, at present anyway, (per @HernanGatta) an OCALL is always dispatched to the host session that did the ECALL that the OCALL is in the context of, so it's straightforward to know _which_ host would apply for an OCALL, which is why it need not be an argument like the enclave is to an ECALL.  However, I think this means the OCALL logic has to be similar to the ECALL logic you show below.

Reply:
The proposal is scoped down to solve the problem in the single-host-multiple-enclaves scenario.